spict/R/utility.R

```diff
@@ -536,3 +536,486 @@ get.cov <- function(rep, parname1, parname2, cor=FALSE){
         return(rep$cov[inds, inds])
     }
 }
+
+
+
+#' @name get.prod.regimes
+#' @title Calculate AIC for spict models with different productivity
+#'     regime shifts
+#' @param inp An input list containing data.
+#' @param years Optional; vector with years for the estimation of the
+#'     regime shift. By default (NULL) all years of the input data are
+#'     used.
+#' @return data frame with years and AIC
+#' @export
+get.prod.regimes <- function(inp, years = NULL){
+    inpin <- check.inp(inp)
+
+    ## set years
+    yearspot <- min(floor(inp$time)):inp$manstart
+    years <- yearspot[yearspot %in% as.numeric(as.character(years))]
+    if(is.null(years) || length(years) == 0) years <- yearspot
+    aiclist <- list()
+
+    ## loop over all years
+    for(yr in years){
+        inp <- inpin
+        inp$MSYregime <- factor(ifelse(inp$time < yr, 1, 2))
+        inp <- check.inp(inp)
+        fit <- try(fit.spict(inp), silent=TRUE)
+        aiclist[[length(aiclist)+1]] <- ifelse(!inherits(fit,"try-error"), get.AIC(fit), NA)
+    }
+    res <- cbind(Year = years, AIC = unlist(aiclist))
+    return(res)
+}
+
+
+#' @name shorten.inp
+#' @title Shorten time series of input data to specified length
+#' @param inp An input list containing data.
+#' @param mintime Minimum time to include in the modified data
+#'     (default: NULL).
+#' @param maxtime Maximum time to include in the modified data
+#'     (default: NULL).
+#' @details Time is given in decimal notation (e.g. 2005.3). If only
+#'     one of the arguments 'minyear' or 'maxyear' is used, the
+#'     function shortens the time series from either end to the
+#'     respective time.
+#' @author T.K. Mildenberger <t.k.mildenberger@gmail.com>
+#' @return Inp list with shortened time series.
+#' @export
+shorten.inp <- function(inp, mintime = NULL, maxtime = NULL){
+    inpin <- check.inp(inp)
+    inpout <- inpin
+
+    ## function to find closest time
+    get.inds <- function(timevec, mintime, maxtime){
+        ## set to limits if NULL or incorrectly set
+        mintimePot <- min(floor(timevec))
+        maxtimePot <- max(ceiling(timevec))
+        if(is.na(mintime) || !is.numeric(mintime) ||
+           mintime < mintimePot || mintime > maxtimePot) mintime <- mintimePot
+        if(is.na(maxtime) || !is.numeric(maxtime) ||
+           maxtime < mintimePot || maxtime > maxtimePot) maxtime <- maxtimePot
+        ## indices
+        ind <- 1:length(timevec)
+        diffmin <- timevec - mintime
```

Comment:
can be in one line? reverse order of timevec - mintine and maxtime - timevec important!